### DISCUSSION TITLE
Backgrounding processes in Android

### DISCUSSION THREAD

**Discussion by hkparker** [fyne-io/fyne#5221] [General]
Fyne packages android apps as a single activity of type [`org.golang.app.GoNativeActivity`](https://github.com/golang/mobile/blob/master/app/GoNativeActivity.java).  Activities cannot run in the background on Android, for that one must use a [Service](https://developer.android.com/develop/background-work/services).  There doesn't appear to be a well carved path to extending Android's service class in go that Fyne could offer right now.  Currently all Fyne apps will stop running a few seconds / minutes after leaving the foreground on Android.  There's been [some discussion](https://stackoverflow.com/a/53888265) on golang android services on stack overflow, but no working examples I could find.

I could imagine, in theory, it being possible to build some go code into a shared object and then start that from Java in an android service implementation.  How Fyne would package that is an open question, but perhaps in the future I could manually build an app that runs Fyne in the activity and my long running code in a service and facilitates communication between the two.  Ideally it would be awesome for Fyne to support a platform-agnostic (sans iOS where this isn't possible) "keep this running" feature, and have representations of background and foreground services.

The app I've created relies on keeping network sockets open in the background at all times for notifications (which is an anti-pattern in mobile development I know, but unavoidable in this case as these sockets are the result of running a Tor client).  So for my use case, there isn't an alternative to backgrounding for the app to function.

**Comment by hkparker** [comment]
Google groups discussion of this [here](https://groups.google.com/g/golang-nuts/c/slHvO4f384g?pli=1).  It does seem like if there was a `GoNativeService` Java class that implemented android's service, and Fyne could hook some code into that in a similar way it does with the `GoNativeActivity`, we could be off to a reasonable start

**Comment by andydotxyz** [comment]
I hope we can deliver a good background service abstraction in 2.6 as we improve the thread handling work in other areas as well.

**Reply by hkparker** [reply]
Cool, yeah I know we've talked about something like this being a goal before, I mostly just wanted to share what I came across for android in the hopes that it's useful later, or in case anyone else runs into the same questions.  I know that cross-platform consistency is an important goal in Fyne, and it looks like this is impossible (or at best unreliable) in iOS, so hopefully that doesn't put a damper on this idea too much.

**Comment by andydotxyz** [comment]
On the call today as we started 2.6 release planning it was agreed this is probably just too much to cram in alongside all the thread work.
Will aim to make it top priority for 2.7

**Reply by hkparker** [reply]
Yeah that makes sense, this would be pretty incredible to see in 2.7!  To be fair I think this is going to be a heavy lift, just from what I've come across while looking into how I'd implement it, but we'll see.  More than happy to help out if I can, this is the most daunting unsolved task in between me and releasing my project.

**Comment by xhlian** [comment]
Hello, any updates or changes to the plan?

**Reply by andydotxyz** [reply]
It remains a priority but we have not yet figured out the API. It falls after the work on Accessibility IMHO.

When there is an update it will be linked here.

**Reply by LeoHalb** [reply]
Hey, I was able to pull together a Fyne app that's using a foreground service on Android.
Essentially it's a Fyne app running as an Android activity (as normal) and a HTTP golang webserver running inside a foreground service, that's handling the long-running logic. The activity and the foreground service are communicating over, you guessed it, HTTP.

I uploaded an example repository here, maybe it helps someone: https://github.com/LeoHalb/FyneForeground

FYI @hkparker as you opened this discussion

**Reply by hkparker** [reply]
Nice, this is awesome!  I'll try to play around with it and build it for my device before too long.  Thank you for sharing this, I'm curious what motivated you to work on it, if you're able to share.

**Reply by hkparker** [reply]
@LeoHalb I was able to build your repo on linux and install the app on my device (pixel 10 / grapheneos).  Works great!  I'd love to talk about the communication between activity and service more.  Are you on the gophers slack, or would you prefer to keep discussions in github issues?

**Reply by LeoHalb** [reply]
@hkparker Hey, glad to hear!
I'm on the gophers slack (as of now), same user handle as this one on GitHub.

**Reply by LeoHalb** [reply]
Update: I got native function calls working.
https://github.com/LeoHalb/FyneForeground/commit/4dcb67d8b9d59fa4cfb22ede0b9c3bcb10e319ce

The underlying problem making this hard is, that you cannot use multiple golang c-shared libraries in the same process: https://github.com/golang/go/issues/65050

To circumvent this issue, I'm starting the Foreground Service in another OS process and instead of communicating over HTTP, I'm using a standard Android way of managing IPC, namely AIDL:
https://medium.com/@anuragsingh7238/aidl-android-interface-definition-language-how-binder-ipc-works-inside-aosp-19465655bd00

At the moment, the code in my repo is heavily AI generated and needs cleanup/refactoring but is working.
I'll first invest some time with these learnings to try and get a [fyne](https://github.com/fyne-io/tools) branch working that is compiling both activity and service into one library, maybe even eliminating the need of IPC altogether...
